lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -0,0 +1,315 @@
+    -> std::set<ClassType> {
+  // find index of super type
+  const auto IndexOfTypeFind = TypeToVertex.find(Type);
+
+  assert(IndexOfTypeFind != TypeToVertex.end());
+
+  size_t IndexOfType = IndexOfTypeFind->getSecond();
+
+  // if the super type hasn't been found, return an empty set
+  if (IndexOfType >= TypeToVertex.size()) {
+    return {};
+  }
+
+  // return all sub types
+  std::set<ClassType> SubTypes = {};
+
+  for (auto Index : TransitiveClosure[IndexOfType].set_bits()) {
+    SubTypes.insert(VertexTypes[Index]);
+  }
+
+  return SubTypes;
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::isSuperType(ClassType Type,
+                                                     ClassType SuperType) {
+  return isSubType(SuperType, Type); // NOLINT
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getSuperTypes(ClassType Type)
+    -> std::set<ClassType> {
+  // TODO: implement (low priority)
+  llvm::report_fatal_error("Not implemented");
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::hasVFTable(ClassType Type) const {
+  const auto TypeIndex = TypeToVertex.find(Type);
+
+  if (TypeIndex == TypeToVertex.end() || TypeIndex->second >= VTables.size()) {
+    return false;
+  }
+
+  return !VTables[TypeIndex->second].empty();
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getVFTable(ClassType Type) const
+    -> const VFTable<f_t> * {
+  const auto TypeIndex = TypeToVertex.find(Type);
+  assert(TypeIndex != TypeToVertex.end());
+  return &VTables[TypeIndex->getSecond()];
+}
+
+void DIBasedTypeHierarchy::print(llvm::raw_ostream &OS) const {
+  size_t VertexIndex = 0;
+  size_t EdgeIndex = 0;
+
+  OS << "Type Hierarchy\n";
+  for (const auto &CurrentVertex : TransitiveClosure) {
+    OS << VertexTypes[VertexIndex]->getName() << "\n";
+    for (size_t I = 0; I < TransitiveClosure.size(); I++) {
+      if (EdgeIndex != VertexIndex && CurrentVertex[I]) {
+        OS << "--> " << VertexTypes[EdgeIndex]->getName() << "\n";
+      }
+      EdgeIndex++;
+    }
+    VertexIndex++;
+    EdgeIndex = 0;
+  }
+
+  OS << "VFTables:\n";
+
+  size_t TypeIndex = 0;
+  for (const auto &Type : VertexTypes) {
+    const auto &VTable = VTables[TypeIndex];
+
+    OS << Type->getName() << ": ";
+
+    // get all function names for the llvm::interleaveComma function
+    llvm::SmallVector<std::string, 6> Names;
+    for (const auto &Function : VTable.getAllFunctions()) {
+      if (Function) {
+        Names.push_back(Function->getName().str());
+      } else {
+        Names.push_back("<none>");
+      }
+    }
+    // prints out all function names, seperated by comma, without a trailing
+    // comma
+    llvm::interleaveComma(Names, OS);
+    OS << "\n";
+    TypeIndex++;
+  }
+
+  OS << "\n";
+}
+
+[[nodiscard]] nlohmann::json DIBasedTypeHierarchy::getAsJson() const {
+  /// TODO: implement
+  llvm::report_fatal_error("Not implemented");
+}
+
+void DIBasedTypeHierarchy::printAsDot(llvm::raw_ostream &OS) const {
+  OS << "digraph TypeHierarchy{\n";
```

Comment:
this would benefit from labels, so that you see them in the graph visualization.

Reply:
What kind of labels? Do you mean like a legend that tells you how the graph is structured?

Reply:
The dot file format allows attributes for nodes and edges, one of which is `label`. There are also more that allow you to set colors, borders etc. when you plot it to a image or just in xdot.

By doing so you could introduce nodes by their id and then reference by id, for example:

`
0[label="Base"]
1[label="Child"]
1->0[label="public"]
`
The label for the edge is just an example. In fact, it might be worth thinking about what would be interesting properties of the edges.

Reply:
I will implement this, thank you